src/modules/programmers/dfu-util.ts

```diff
@@ -17,46 +20,114 @@ if (process.platform === 'win32') {
 
 log.info(`dfuUtil found at ${dfuUtil}`);
 
-export function stm32(): void {
-  const firmware = window.inputPath;
-  console.log(firmware);
-  if (firmware.endsWith('bin')) {
-    const command = dfuUtil;
-    const args = [
-      '-a',
-      '0',
-      '-d',
-      '0483:df11',
-      '-s',
-      '0x08000000:leave',
-      '-D',
-      firmware,
-    ];
-    try {
-      const dfuer = spawn(command, args);
-      dfuer.stdout.on('data', window.Bridge.statusAppendNoLF);
-      dfuer.stderr.on('data', window.Bridge.statusAppendNoLF);
-    } catch (error) {
-      window.Bridge.statusAppend(` error: ${error}`);
-    }
-  } else {
-    window.Bridge.statusAppend(` Err: dfu-util only works with .bin files`);
-  }
+export function flasher(
+  command: string,
+  args: Array<string>
+): Promise<boolean | Error> {
+  return new Promise((resolve, reject) => {
+    const dfuer = spawn(command, args);
+    dfuer.stdout.on('data', window.Bridge.statusAppendNoLF);
+    dfuer.stderr.on('data', window.Bridge.statusAppendNoLF);
+
+    const cancelID = timeoutBuilder(reject, dfuer, 'Flash Timedout', 15000);
+    dfuer.on('exit', (code: unknown /*, signal*/) => {
+      clearTimeout(cancelID);
+      if (code === 0) {
+        resolve(true);
+      } else {
+        if (code === null) {
+          reject(new Error('Flash Timedout'));
+        } else {
+          reject(new Error(`Flash Failed ${code}`));
+        }
+      }
+    });
+  }) as Promise<boolean | Error>;
+}
+
+export function stm32(firmware: string): Promise<boolean | Error> {
+  const command = dfuUtil;
+  const args = [
+    '-a',
+    '0',
+    '-d',
+    '0483:df11',
+    '-s',
+    '0x08000000:leave',
+    '-D',
+    firmware,
+  ];
+  return flasher(command, args) as Promise<boolean | Error>;
 }
 
-export function kiibohd(): void {
-  const f = window.inputPath;
-  if (f.endsWith('bin')) {
-    const command = dfuUtil;
-    const args = ['-D', f];
-    try {
-      const dfuer = spawn(command, args);
-      dfuer.stdout.on('data', window.Bridge.statusAppendNoLF);
-      dfuer.stderr.on('data', window.Bridge.statusAppendNoLF);
-    } catch (error) {
-      window.Bridge.statusAppend(` error: ${error}`);
-    }
-  } else {
-    window.Bridge.statusAppend(` Err: dfu-util only works with .bin files`);
+export function kiibohd(filename: string): Promise<boolean | Error> {
+  const command = dfuUtil;
+  const args = ['-D', filename];
+  return flasher(command, args) as Promise<boolean | Error>;
+}
+
+export enum Family {
+  stm32,
+  kiibohd,
+}
+
+export class DFUUtil {
+  constructor(filename: string, family: Family) {
+    this.filename = filename;
+    this.family = family;
+  }
```

Comment:
Like your doing with dfu programmer should we not also accept the status append tools here as well?

Reply:
yep, it's a work in progress. Now that we have the usb-detection stuff working again I will probably put some time in this weekend as we have a 4 day weekend coming up. Thanks for review @e11i0t23 appreciated.